hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/listener/SharedTopicListener.java

```
@@ -40,19 +47,61 @@
+        Flux<TopicMessage> topicMessageFlux = getSharedListener(filter)
                 .doOnSubscribe(s -> log.info("Subscribing: {}", filter))
-                .doOnCancel(() -> processor.onNext("timeout"))
-                .onBackpressureBuffer(listenerProperties.getMaxBufferSize())
-                .timeout(timeoutFlux, message -> timeoutFlux, Mono.error(
-                        new ClientTimeoutException("Client timed out while consuming the buffered messages")));
+                .onBackpressureBuffer(listenerProperties.getMaxBufferSize(), t -> timeoutContext.onOverflow())
+                .doFinally(r -> timeoutContext.onComplete());
+        return Flux.merge(listenerProperties.getPrefetch(), topicMessageFlux, timeoutMono)
+                .publishOn(Schedulers.boundedElastic(), false, listenerProperties.getPrefetch());
     }
 
     protected abstract Flux<TopicMessage> getSharedListener(TopicMessageFilter filter);
+
+    @RequiredArgsConstructor
+    private class TimeoutContext implements Consumer<MonoSink<TopicMessage>> {
+
+        private final Scheduler scheduler;
+        private final Duration timeout;
+
+        private MonoSink<TopicMessage> sink;
+        private final AtomicBoolean completed = new AtomicBoolean(false);
+        private volatile Disposable taskDisposer;
+
+        @Override
+        public void accept(MonoSink<TopicMessage> sink) {
+            this.sink = sink;
+        }
+
+        public void onComplete() {
+            if (!completed.compareAndSet(false, true)) {
+                return;
+            }
+
+            if (taskDisposer != null) {
+                taskDisposer.dispose();
+                taskDisposer = null;
+            }
+
+            sink.success();
+        }
+
+        public void onOverflow() {
+            if (completed.get()) {
+                return;
+            }
+
+            try {
+                taskDisposer = scheduler.schedule(() -> sink.error(new ClientTimeoutException(
```

As discussed offline, I think we can remove the buffer timeout and just fail fast. This will simplify the logic.

---

removed, now will fail fast